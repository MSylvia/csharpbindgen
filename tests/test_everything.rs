use insta::assert_snapshot_matches;

use csharpbindgen::{
    Builder,
    CSAccess
};

#[test]
fn test_it_works() {
    let rust_code = r#"
        pub const BOOP: u8 = 1;
        pub const IGNORE_THIS_CONST: u8 = 2;

        #[repr(C)]
        pub struct MyStruct {
            pub foo: f32,
            pub bar: f32
        }

        #[repr(C)]
        pub struct IgnoreThisStruct { pub foo: f32 }

        pub type MyOpaqueRef = *mut MyOpaqueStruct;

        pub unsafe extern "C" fn public_func() {}

        #[repr(C)]
        pub struct PublicStruct {
            pub bop: i32,
        }

        fn unexported_func() {}

        pub unsafe extern "C" fn blarg(
            a: i32,
            b: *const MyStruct,
            c: MyOpaqueRef
        ) -> u8 { 120 }

        pub unsafe extern "C" fn ignore_this_func(a: i32) -> u8 { 120 }
    "#;

    let code = Builder::new("MyDll", String::from(rust_code))
        .class_name("MyStuff")
        .ignore(&["ignore_*", "IGNORE_*", "Ignore*"])
        .access("public_func", CSAccess::Public)
        .access("PublicStruct", CSAccess::Public)
        .generate()
        .unwrap();

    assert_snapshot_matches!("main_example", code);
}
