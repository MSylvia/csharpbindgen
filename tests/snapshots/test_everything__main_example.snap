---
created: "2019-11-23T12:03:30.755697800Z"
creator: insta@0.8.1
source: tests/test_everything.rs
expression: code
---
// This file has been auto-generated, please do not edit it.

using System;
using System.Runtime.InteropServices;

[Serializable]
[StructLayout(LayoutKind.Sequential)]
internal struct MyStruct {
    internal Single foo;
    internal Single bar;

    internal MyStruct(Single foo, Single bar) {
        this.foo = foo;
        this.bar = bar;
    }
}

[Serializable]
[StructLayout(LayoutKind.Sequential)]
public struct PublicStruct {
    public Int32 bop;

    public PublicStruct(Int32 bop) {
        this.bop = bop;
    }
}

internal class MyStuff {
    internal const Byte BOOP = 1;

    [DllImport("MyDll")]
    public static extern void public_func();

    [DllImport("MyDll")]
    internal static extern Byte blarg(Int32 a, ref MyStruct b, IntPtr /* MyOpaqueStruct */ c);

}

